org.jacoco.core/src/org/jacoco/core/instr/Instrumenter.java

```diff
@@ -200,6 +203,12 @@ public int instrumentAll(final InputStream input,
 
 	private int instrumentZip(final InputStream input,
 			final OutputStream output, final String name) throws IOException {
+		if (generators.getLast() instanceof
+				OfflineInstrumentationCompanionAccessGenerator) {
+			generators
+					.addLast(new
+							OfflineInstrumentationCompanionAccessGenerator());
```

Comment:
If the general policy is objects are thread-unsafe unless otherwise specified, then doing nothing is fine with me.  There's no API "break" since client code will need to opt into using OfflineInstrumentationCompanionAccessGenerator anyway.

Creating a new Instrumenter per nested archive seems cleaner than using a LinkedList.  If you want to allow clients to extend Instrumenter, you could add an Instrumenter(Instrumenter) copy constructor and a newInstrumenter(Instrumenter) method that clients could override.  Again, nothing would change for current clients since that method would be unused unless the client opted in to using OfflineInstrumentationCompanionAccessGenerator.  If you'd rather keep the LinkedList mechanism, then perhaps move the companionGenerator local variable to the outer block rather than doing `generators.getLast() instanceof ...` twice?